snowfakery/template_funcs.py

```diff
@@ -130,14 +130,26 @@ def reference(self, x: Union[ObjectRow, str]):
             if hasattr(x, "id"):  # reference to an object with an id
                 target = x
             elif isinstance(x, str):  # name of an object
-                obj = self.context.field_vars().get(x)
-                if not obj:
+                # allows dotted paths
+                parts = x.split(".")
+                target = self.context.field_vars().get(parts.pop(0))
+
+                for part in parts:
+                    try:
+                        target = getattr(target, part)
```

Comment:
We should make sure it's not possible to traverse to dunder attributes this way.

Reply:
It probably is but it always was in Jinja. I don't think we should think of Snowfakery files as being any more secure than Python code (at least as long as we run them in a Python runtime wth Jinja enabled).

Traversing through leadiing-double-underscore is probably useful because Snowfakery uses them but I could disallow traversing through trailing-double-underscore identifiers.

Using Jinja:

```Bar(id=1, bar=<bound method ObjectRow.__getattr__ of <ObjectRow Foo 1>>)```

I should probably add a note to the docs about only executing trusted recipes or recipes in a hardened context like a container.

Reply:
For jinja2 we can and should be using the sandboxed environment: https://jinja.palletsprojects.com/en/2.11.x/sandbox/

And if we don't do that, then yes, we should make it a lot clearer that running a snowfakery recipe is equivalent to running code.

Reply:
Unfortunately it doesn't seem to be possible to simultaneously get the benefits of both the NativeEnvironment and the SandboxEnvironment and I have a PR open to use the NativeEnvironment.

#185

Reply:
Okay. Waiting for you to add that clear warning to the docs then.

Reply:
8a14ebf37f26d24eea3eb429ba285c6fe0d99561